dhcp-controller/dhcpserver/pkg/server/server.go

```diff
@@ -0,0 +1,400 @@
+				}
+				// This loop only runs once
+				for ip, obj := range ipallocation.Spec.Allocations {
+					tmpline := fmt.Sprintf(ipallocation.Spec.EpochExpiry + " " + obj.MacId + " " + ip + " " + obj.VmiRef + " *\n")
+					destination.WriteString(tmpline)
+				}
+				serverLog.Info("Restored IPAllocation " + ipallocation.Name)
+			}
+		}
+
+	}
+
+	return err
+}
+
+// Check if the lease file exists
+func dirInit(leasePath string) {
+	_, err := os.Stat(leasePath)
+	if os.IsNotExist(err) {
+		// If not, retrieve from backup and create leasefile
+		retrieveBackup(leasePath)
+	}
+}
+
+// Checks if lease found in records exist in leasefile.
+// This is for the scenario when leases expire and dnsmasq deletes
+// the lease from the leasefile
+func leaseExist(ip string, records [][]string) (result bool) {
+	result = false
+	for _, lease := range records {
+		if lease[2] == ip {
+			result = true
+			break
+		}
+	}
+	return result
+}
+
+// Checks if existing entry in leasefile has been updated
+func checkRecord(lease LeaseFile, record []string) bool {
+	serverLog.Info("Checking lease entry....")
+	var isupdated = false
+	if record[0] != lease.EpochTimestamp ||
+		record[1] != lease.MacAddress ||
+		record[2] != lease.IPAddress ||
+		record[3] != lease.Hostname ||
+		record[4] != lease.ClientID {
+		isupdated = true
+	}
+	return isupdated
+}
+
+// updates record with new data from leasefile
+func updateRecord(lf map[string]LeaseFile, record []string, isupdate bool) {
+	serverLog.Info("Updating Lease")
+	lf[record[2]] = LeaseFile{record[0], record[1], record[2], record[3], record[4]}
+
+	if isupdate {
+		_, err := k8sClient.UpdateIPAllocation(context.TODO(), record[0], record[1], record[3], record[2])
+		if err != nil {
+			serverLog.Error(err, "failed to update IP allocation")
+		}
+	} else {
+		_, err := k8sClient.CreateIPAllocation(context.TODO(), record[0], record[1], record[3], record[2])
+		if err != nil {
+			serverLog.Error(err, "failed to create IP allocation")
+		}
+	}
+
+}
+
+func readLeaseFile(lf map[string]LeaseFile, leasePath string) (string, error) {
+	serverLog.Info("Reading leasefile...")
+	f, err := os.Open(leasePath)
+	if err != nil {
+		serverLog.Error(err, "Cannot open lease file")
+		return "", err
+	}
+	defer f.Close()
+
+	csvReader := csv.NewReader(f)
+	csvReader.Comma = ' '
+	records, _ := csvReader.ReadAll()
+	if err != nil {
+		serverLog.Error(err, "Cannot parse lease file")
+		return "", err
+	}
+
+	// Check if record exists in leasefile
+	for ip, _ := range lf {
+		if !leaseExist(ip, records) {
+			delete(lf, ip)
+			isdeleted, err := k8sClient.DeleteIPAllocation(context.TODO(), ip)
+			if err != nil {
+				serverLog.Error(err, "failed to delete IP allocation")
+			}
+			if isdeleted {
+				serverLog.Info("Deleted IPAllocation " + ip)
+			}
+		}
+	}
+
+	// Check if lease exists in record and if it is up to date
+	for _, record := range records {
+		if lease, ok := lf[record[2]]; ok {
```

Comment:
Same as above, see if we can change it into a structured format or map[string] as it isnt clear what the format of a record is or what data it holds

Reply:
Done. Changed it to a typed struct.